src/main/java/com/amazonaws/services/dynamodbv2/datamodeling/internal/LRUCache.java

```diff
@@ -98,11 +99,16 @@ public void clear() {
         // The more complicated logic is to ensure that the listener is
         // actually called for all entries.
         if (listener != null) {
-            Set<String> keys = new TreeSet<String>(map.keySet());
-            for (String key : keys) {
-                T val = map.get(key);
-                listener.onRemoval(new SimpleImmutableEntry<String, T>(key, val));
-                map.remove(key);
+            List<Entry<String, T>> removedEntries = new ArrayList<Entry<String, T>>();
+            synchronized (map) {
+                Iterator<Entry<String, T>> it = map.entrySet().iterator();
+                while(it.hasNext()) {
+                    removedEntries.add(it.next());
+                    it.remove();
+                }
+            }
+            for (Entry<String, T> entry : removedEntries) {
+                listener.onRemoval(entry);
```

Comment:
Does anything depend on the synchronization we're removing here? It's now possible to have onRemove called for the same key K twice, with a newer version being called before the older one.

Reply:
Ah, never mind, I thought it was moving the onRemoval outside of the synchronized block, but now I see there wasn't synchronization before.